Add Outputter with Formatter

ronan opened the pull request:
Here's a WIP on adding a class that allows all returned values to be output in the same way. 

I added 3 new interfaces: 
- _OutputFormatter_ - Formats a return value (eg: JSON, Pretty tables)
- _OutputWriter_ - Writes the output somewhere (eg: STDOUT)
- _Outputter_ - Takes a formatter and a writer and does the magic

I implemented a stream writer and a JSON formatter. Bash and human-readable formatters are still needed.

I also changed the 'sites list' function to use the outputter instead of formatting and outputting it's return value directly.

I can continue converting commands to use this outputter and work on implementing the other formatters if this looks like good direction.

Commit:
Switched Terminus logger for katzgrau/KLogger

Commit:
Adds basis for outputter classes

Commit:
Fixes tests for outputters

Commit:
Fixes outputters

Commit:
Adds output formatting to 'sites list'

SaraMcCutcheon commented on php/Terminus/Outputters/Outputter.php:
```diff
@@ -0,0 +1,119 @@
+
+  /**
+   * @param \Terminus\Outputters\OutputWriterInterface $writer
+   * @param \Terminus\Outputters\OutputFormatterInterface $formatter
```
Namespace is not global here (i.e. not beginning with '/')

Commit:
Adds human-readable formatter

Commit:
Converts all handleDisplay calls to the output formatter

Commit:
Merge branch 'master' into outputter

Conflicts:
	composer.lock
	php/Terminus/KLogger.php
	php/Terminus/Runner.php
	php/class-terminus.php

ronan commented:
@TeslaDethray I've added a human-readable 'pretty' formatter which uses the ascii table from the original output handler. I've also converted every instance of handleDisplay to use the output formatter. 

There are some TODOs in the cli command because some of those output json only. Switching to the new formatters would make pretty-printing the default (with the --json and --bash flags available). I think this is appropriate but it changes the current default behavior and may break scripts, so I wanted to get your take first.

I still need to create BASH formatter. I'll try and get to that tonight.

SaraMcCutcheon commented:
This looks great! Such an improvement! Once the CI passes (with no major changes, of course), I +1 it.

Commit:
Adds bash output formatter

Commit:
Merge branch 'master' into outputter

Conflicts:
	composer.json
	composer.lock

Commit:
Adds dump formatting and uses outputter for cli commands

ronan commented:
Ok, Bash formatter added. I tried to copy the current behavior.

I also converted the cli dump functions to use the outputter. 

Pretty much everything that goes to stdout now should go through the outputter. That means that the format flags should be pretty trustable for scripting (and there won't be any need to parse out warnings and info strings).

As far as I can tell the only parts of the code that access stdout without going through the outputter now are the help (shouldn't be needed in automation) and wp-cli/drush. We could capture the drush output and repackage it but that would be out of scope for this pull and isn't necessary since --format=json should work. Don't know about wp-cli but that's out of scope here too.

Commit:
Merge branch 'master' into outputter

Commit:
Adds bash and pretty format tests

Commit:
Improves value flattening for output

Commit:
Fixes bad merge of update dependencies

Commit:
Removes invalid assertion from JSON test

Commit:
Fixes phpdoc for outputters

Commit:
Fixes strict warning with non-static function

ronan commented:
Current test failures are due to my 'prettification' of the output for the default (human readable) output format. We can

1) Change the tests to look for humanized labels
2) Change the tests to ask for a machine format (--bash or --json)
3) Change the default format to include the machine labels

Thoughts?

SaraMcCutcheon commented:
@ronan We should definitely change the expected test outputs to match the outputter. We've not been using the Behat tests as contract yet, but that should be coming soon and your prettified outputs are _definitely_ what I want to see in there when that happens. (Definitely.)

Commit:
Changed tests to conform to new outputter

SaraMcCutcheon commented:
I've changed the tests. They should pass CI and this can now be built upon.

SaraMcCutcheon commented:
This change also closes #197

fujin commented:
Code and tests look good here. I'm not sure if there are any style violations per your policies (will the travis PR do those?), but LGTM :+1:

SaraMcCutcheon commented:
Travis does not yet check thoroughly for linting, as there are still too many files that do not yet conform, but we've been checking them along the way. I expect it will be fully linted before 1.0.0.
